source/mir/combinatorics/package.d

```diff
@@ -0,0 +1,373 @@
+{
+    static if (isFloatingPoint!R)
+    {
+        import std.mathspecial: gamma;
+        return gamma(n+1);
+    }
+    else static if (is(R == uint) || is(R == int))
+    {
+        // all integer values (0-12)
+        if(0 <= n && n < 13)
+        {
+            static immutable res = [1, 1, 2, 6, 24, 120, 720, 5040, 40320,
+                362880, 3628800, 39916800, 479001600];
+            return res[n];
+        }
+    }
+    else static if (is(R == ulong) || is(R == long))
+    {
+        // all long values 0-20
+        if (0 <= n && n < 21)
+        {
+            static immutable res = [1, 1, 2, 6, 24, 120, 720, 5040, 40320,
+                362880, 3628800, 39916800, 479001600, 6227020800, 87178291200,
+                1307674368000, 20922789888000, 355687428096000, 6402373705728000,
+                121645100408832000, 2432902008176640000];
+            return res[n];
+        }
+    }
+    else
+    {
+        if(n > 19)
+        {
+            R result = factorial!ulong(20);
+            T target = n + 1;
+            for (T i = 21; i < target;i++)
+            {
+                result *= i;
+            }
+            return result;
+        }
+        else if(n >= 0)
+        {
+            R result = factorial!ulong(n);
+            return result;
+        }
+    }
+    //assert(0, "negative numbers are not defined");
+    static if (!isFloatingPoint!R)
+    {
+        R result = 0;
+        return result;
+    }
+}
+
+// convenience overload if target has the same type
+/// ditto
+auto factorial(T)(T n)
+{
+    return factorial!(T, T)(n);
+}
+
+unittest
+{
+    assert(factorial(0) == 1);
+    assert(factorial(1) == 1);
+    assert(factorial(2) == 2);
+    assert(factorial(10) == 3628800);
+    assert(factorial!ulong(20) == 2432902008176640000);
+
+    import std.bigint;
+    assert(factorial!BigInt(30) == BigInt("265252859812191058636308480000000"));
+}
+
+unittest
+{
+    import std.bigint;
+    assert(factorial!BigInt(0) == 1);
+    assert(factorial!BigInt(1) == 1);
+    assert(factorial!BigInt(10) == 3628800);
+
+    // floating point
+    import std.math: approxEqual;
+    assert(factorial(5.5).approxEqual(287.885, 0.001));
+    assert(factorial!float(5.5).approxEqual(287.885, 0.001));
+    assert(factorial!real(5.5).approxEqual(287.885, 0.001));
+}
+
+struct IndexedRoR(Collection, Range)
+if (isInputRange!Range)
+{
+    import std.range : indexed, isForwardRange;
+
+    private Collection c;
+    private Range r;
+
+    this(Collection c, Range r)
+    {
+        this.c = c;
+        this.r = r;
+    }
+
+    @property auto ref front()
+    {
+        return r.indexed(c.front);
+    }
+
+    void popFront()
+    {
+        c.popFront;
+    }
+
+    @property bool empty()
+    {
+        return c.empty;
+    }
+
+    @property size_t length()
+    {
+        return c.length;
+    }
+
+    static if(isForwardRange!Range)
+    {
+        @property typeof(this) save()
+        {
+            typeof(this) t = this;
```

Comment:
this gives `constructor call must be in a constructor` - but we can call IndexedRoR ;-)